File: sidecar/src/services/ScrapbookStorage.ts
/**
 * ScrapbookStorage.ts
 * 
 * Service for managing scrapbook requests (API Explorer quick requests).
 * Stores requests in ~/.apinox/scrapbook.json for persistence across sessions.
 */

import * as fs from 'fs';
import * as path from 'path';
import { ScrapbookRequest, ScrapbookState } from '../../../shared/src/models';
import { DiagnosticService } from './DiagnosticService';

export class ScrapbookStorage {
    private scrapbookPath: string;

    constructor() {
        const homeDir = process.env.HOME || process.env.USERPROFILE || '';
        const configDir = path.join(homeDir, '.apinox');
        
        // Ensure config directory exists
        if (!fs.existsSync(configDir)) {
            fs.mkdirSync(configDir, { recursive: true });
        }
        
        this.scrapbookPath = path.join(configDir, 'scrapbook.json');
        this.log(`ScrapbookStorage initialized: ${this.scrapbookPath}`);
    }

    private log(message: string) {
        DiagnosticService.getInstance().log('BACKEND', `[ScrapbookStorage] ${message}`);
    }

    /**
     * Load scrapbook from disk
     */
    public loadScrapbook(): ScrapbookState {
        try {
            if (!fs.existsSync(this.scrapbookPath)) {
                this.log('Scrapbook file does not exist, returning empty state');
                return { requests: [] };
            }

            const content = fs.readFileSync(this.scrapbookPath, 'utf8');
            const state = JSON.parse(content) as ScrapbookState;
            
            this.log(`Loaded scrapbook with ${state.requests.length} request(s)`);
            return state;
        } catch (error: any) {
            this.log(`Error loading scrapbook: ${error.message}. Returning empty state.`);
            return { requests: [] };
        }
    }

    /**
     * Save scrapbook to disk
     */
    public saveScrapbook(state: ScrapbookState): void {
        try {
            const content = JSON.stringify(state, null, 2);
            fs.writeFileSync(this.scrapbookPath, content, 'utf8');
            this.log(`Saved scrapbook with ${state.requests.length} request(s)`);
        } catch (error: any) {
            this.log(`Error saving scrapbook: ${error.message}`);
            throw new Error(`Failed to save scrapbook: ${error.message}`);
        }
    }

    /**
     * Add a new request to scrapbook
     */
    public addRequest(request: ScrapbookRequest): ScrapbookState {
        const state = this.loadScrapbook();
        
        // Ensure timestamps are set
        const now = new Date().toISOString();
        const newRequest: ScrapbookRequest = {
            ...request,
            createdAt: request.createdAt || now,
            lastModified: now
        };
        
        state.requests.push(newRequest);
        this.saveScrapbook(state);
        
        this.log(`Added request: ${newRequest.name} (${newRequest.id})`);
        return state;
    }

    /**
     * Update an existing request
     */
    public updateRequest(id: string, updates: Partial<ScrapbookRequest>): ScrapbookState {
        const state = this.loadScrapbook();
        const index = state.requests.findIndex(r => r.id === id);
        
        if (index === -1) {
            throw new Error(`Request with id ${id} not found`);
        }
        
        // Update request and set lastModified
        state.requests[index] = {
            ...state.requests[index],
            ...updates,
            id, // Preserve ID
            lastModified: new Date().toISOString()
        };
        
        this.saveScrapbook(state);
        this.log(`Updated request: ${state.requests[index].name} (${id})`);
        return state;
    }

    /**
     * Delete a request
     */
    public deleteRequest(id: string): ScrapbookState {
        const state = this.loadScrapbook();
        const index = state.requests.findIndex(r => r.id === id);
        
        if (index === -1) {
            throw new Error(`Request with id ${id} not found`);
        }
        
        const deletedRequest = state.requests[index];
        state.requests.splice(index, 1);
        this.saveScrapbook(state);
        
        this.log(`Deleted request: ${deletedRequest.name} (${id})`);
        return state;
    }

    /**
     * Get a specific request by ID
     */
    public getRequest(id: string): ScrapbookRequest | null {
        const state = this.loadScrapbook();
        const request = state.requests.find(r => r.id === id);
        return request || null;
    }

    /**
     * Get all requests
     */
    public getAllRequests(): ScrapbookRequest[] {
        const state = this.loadScrapbook();
        return state.requests;
    }
}

File: src-tauri/src/scrapbook_storage.rs
use std::fs;
use std::path::PathBuf;

/// Scrapbook state — stores requests as raw JSON to preserve all frontend fields
#[derive(Debug, serde::Serialize, serde::Deserialize, Default)]
struct ScrapbookState {
    requests: Vec<serde_json::Value>,
}

/// Get path to scrapbook.json file
fn get_scrapbook_path() -> Result<PathBuf, String> {
    let config_dir = std::env::var("APINOX_CONFIG_DIR")
        .ok()
        .and_then(|dir| if dir.trim().is_empty() { None } else { Some(PathBuf::from(dir)) })
        .or_else(|| {
            let home = std::env::var("HOME")
                .or_else(|_| std::env::var("USERPROFILE"))
                .ok()?;
            Some(PathBuf::from(home).join(".apinox"))
        })
        .ok_or("Could not determine config directory")?;

    // Ensure directory exists
    if !config_dir.exists() {
        fs::create_dir_all(&config_dir)
            .map_err(|e| format!("Failed to create config directory: {}", e))?;
    }

    Ok(config_dir.join("scrapbook.json"))
}

/// Load scrapbook from disk
fn load_scrapbook(scrapbook_path: &PathBuf) -> ScrapbookState {
    if !scrapbook_path.exists() {
        return ScrapbookState::default();
    }

    match fs::read_to_string(scrapbook_path) {
        Ok(content) => match serde_json::from_str(&content) {
            Ok(data) => data,
            Err(e) => {
                log::error!("Failed to parse scrapbook file: {}. Starting with empty scrapbook.", e);
                ScrapbookState::default()
            }
        },
        Err(e) => {
            log::error!("Failed to read scrapbook file: {}. Starting with empty scrapbook.", e);
            ScrapbookState::default()
        }
    }
}

/// Save scrapbook to disk
fn save_scrapbook(scrapbook_path: &PathBuf, data: &ScrapbookState) -> Result<(), String> {
    let content = serde_json::to_string_pretty(data)
        .map_err(|e| format!("Failed to serialize scrapbook: {}", e))?;
    
    fs::write(scrapbook_path, content)
        .map_err(|e| format!("Failed to write scrapbook file: {}", e))?;
    
    log::info!("Saved scrapbook with {} request(s)", data.requests.len());
    Ok(())
}

/// Get all scrapbook requests
#[tauri::command]
pub async fn get_scrapbook() -> Result<Vec<serde_json::Value>, String> {
    let scrapbook_path = get_scrapbook_path()?;
    let data = load_scrapbook(&scrapbook_path);
    Ok(data.requests)
}

/// Add a new request to scrapbook
#[tauri::command]
pub async fn add_scrapbook_request(request: serde_json::Value) -> Result<Vec<serde_json::Value>, String> {
    let scrapbook_path = get_scrapbook_path()?;
    let mut data = load_scrapbook(&scrapbook_path);

    let mut req = request;
    let now = chrono::Utc::now().to_rfc3339();
    if req.get("createdAt").is_none() {
        req["createdAt"] = serde_json::Value::String(now.clone());
    }
    req["lastModified"] = serde_json::Value::String(now);

    data.requests.push(req);
    save_scrapbook(&scrapbook_path, &data)?;
    Ok(data.requests)
}

/// Update an existing scrapbook request (merges fields)
#[tauri::command]
pub async fn update_scrapbook_request(
    id: String,
    updates: serde_json::Value,
) -> Result<Vec<serde_json::Value>, String> {
    let scrapbook_path = get_scrapbook_path()?;
    let mut data = load_scrapbook(&scrapbook_path);

    let req = data.requests.iter_mut()
        .find(|r| r.get("id").and_then(|v| v.as_str()) == Some(&id))
        .ok_or_else(|| format!("Request with id {} not found", id))?;

    if let (serde_json::Value::Object(req_map), serde_json::Value::Object(updates_map)) = (req, &updates) {
        for (k, v) in updates_map {
            req_map.insert(k.clone(), v.clone());
        }
        req_map.insert("lastModified".to_string(), serde_json::Value::String(chrono::Utc::now().to_rfc3339()));
    }

    save_scrapbook(&scrapbook_path, &data)?;
    Ok(data.requests)
}

/// Delete a scrapbook request
#[tauri::command]
pub async fn delete_scrapbook_request(id: String) -> Result<Vec<serde_json::Value>, String> {
    let scrapbook_path = get_scrapbook_path()?;
    let mut data = load_scrapbook(&scrapbook_path);

    data.requests.retain(|r| r.get("id").and_then(|v| v.as_str()) != Some(&id));
    save_scrapbook(&scrapbook_path, &data)?;
    Ok(data.requests)
}

/// Get a specific scrapbook request by ID
#[tauri::command]
pub async fn get_scrapbook_request(id: String) -> Result<Option<serde_json::Value>, String> {
    let scrapbook_path = get_scrapbook_path()?;
    let data = load_scrapbook(&scrapbook_path);
    Ok(data.requests.into_iter().find(|r| r.get("id").and_then(|v| v.as_str()) == Some(&id)))
}
